docs/proposals/post-deployment-hooks.md

```diff
@@ -0,0 +1,187 @@
+
+Post-deployment hooks implemented external to the Kubelet provide a different set of guarantees:
+
+1. Hooks can be bound to the logical deployment lifecycle, enabling hook executions decoupled from replication mechanics.
+  1. For example, races can be avoided by defining a hook that executes exactly once per deployment regardless of the replica count.
+2. Hooks defined in terms of deployments are conceptually easier to reason about from the perspective of a user defining a deployment workflow.
+
+There are special considerations imposed by such an external hook mechanism:
+
+1. The hooks are executed asynchronous relative to the pod lifecycle.
+2. The hook outcomes have no direct impact on pod lifecycle.
+
+These considerations introduce disadvantages in some use cases:
+
+1. Nothing prevents external connectivity races: when a pod's containers are created, they become routable to services and other pods within an application the moment the containers enter a ready state, likely before or during hook execution.
+2. Hook execution can't be atomically linked to the pod's status: there will be a window where the pod is in a ready state, and a failed hook execution must revert the pod's state back to failed when it would be preferable to have never allowed to the pod to become ready prior to hook success.
+
+## Use cases
+
+1. As a MySQL user, I want my application deployment to seed the database with data the first time the database is used.
+2. As a MySQL user, I want slave instance deployments to be automatically registered with the master.
+3. As a MongoDB user, I want database peer instances to be automatically clustered during deployments.
+4. As a Rails application author, I want to perform a Rails database migration following an application deployment.
+
+#### Use-case: Seeding a MySQL database
+
+Databases are often initialized with example or other data the first time they are deployed. Users should be able to express a hook with causes the database to be initialized automatically.
+
+Consider an application deployment which uses a MySQL database via a service. In this use case, it's the responsibility of the application deployment to initialize the database.
+
+###### Container lifecycle hooks
+
+A container lifecycle hook associated with the application deployment suffers from race conditions: the hook would execute once per pod in the deployment, burdening the hook author to ensure the hook is idempotent and safely concurrent when faced with races to initialize the database.
+
+###### Post-deployment hooks
+
+A post-deployment hook could satisfy the use case by ensuring the initialization is performed only once per logical deployment.
+
+#### Use-case: Clustering a MySQL database
+
+MySQL requires an orchestrated workflow for slave cluster participation. For a slave to join the cluster, the master must be locked and exported, and slaves associated with the master. Users should be able to deploy clustered MySQL instances.
```

Comment:
>This may be a bad example, because slaves and masters are probably in different deployment configs. Also, this sort of configuration is the first setup, rather than on every code release.

Hm. Does this mean that post-start hooks are actually more appropriate? You want the registration to occur every time a replica comes into existence, not as a side effect of a deployment. Seems to relate more to scaling than deployment.